File: projects/chess-coach/apps/harness/src/main/kotlin/harness/StateManager.kt
package harness

import java.io.File

class StateManager(
    private val fenFilePath: String = resolveStatePath("game_state.fen")
) {
    fun readFen(): String {
        val file = File(fenFilePath)
        return if (file.exists()) {
            file.readText().trim()
        } else {
            // Default starting position
            "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"
        }
    }

    fun writeFen(fen: String) {
        File(fenFilePath).parentFile?.mkdirs()
        File(fenFilePath).writeText(fen)
    }

    private companion object {
        fun resolveStatePath(fileName: String): String {
            val dir = System.getenv("CHESS_STATE_DIR")?.trim().orEmpty()
            return if (dir.isEmpty()) fileName else "$dir/$fileName"
        }
    }
}

File: projects/chess-coach/apps/server/src/StateManager.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";

const STARTING_FEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";

function resolveStatePath(fileName: string): string {
  const dir = (process.env.CHESS_STATE_DIR ?? "").trim();
  return dir === "" ? fileName : `${dir}/${fileName}`;
}

/**
 * File-backed FEN state, mirroring the Kotlin StateManager + StateReader
 * (_spec/api/behavior.md §8). Honors CHESS_STATE_DIR.
 */
export class StateManager {
  readonly startingFen = STARTING_FEN;

  constructor(private readonly fenFilePath: string = resolveStatePath("game_state.fen")) {}

  readFen(): string {
    if (existsSync(this.fenFilePath)) {
      return readFileSync(this.fenFilePath, "utf8").trim();
    }
    return STARTING_FEN;
  }

  writeFen(fen: string): void {
    const dir = dirname(this.fenFilePath);
    if (dir && dir !== "." && !existsSync(dir)) mkdirSync(dir, { recursive: true });
    writeFileSync(this.fenFilePath, fen);
  }

  resetGame(): void {
    this.writeFen(STARTING_FEN);
  }
}
